### PR TITLE
Fix exception when ClassSelectors contain interfaces

### LINKED ISSUE #1: ClassSelectors for interfaces should be ignored
When a `ClassSelector` for an interface type is passed as part of a discovery request, discovery currently fails with an exception:

```
Caused by: org.junit.platform.commons.JUnitException: TestEngine with ID 'testng' failed to discover tests
	at org.junit.platform.launcher.core.EngineDiscoveryOrchestrator.discoverEngineRoot(EngineDiscoveryOrchestrator.java:111)
	at org.junit.platform.launcher.core.EngineDiscoveryOrchestrator.discover(EngineDiscoveryOrchestrator.java:85)
	at org.junit.platform.launcher.core.DefaultLauncher.discover(DefaultLauncher.java:92)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:75)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor$CollectAllTestClassesExecutor.processAllTestClasses(JUnitPlatformTestClassProcessor.java:99)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor$CollectAllTestClassesExecutor.access$000(JUnitPlatformTestClassProcessor.java:79)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor.stop(JUnitPlatformTestClassProcessor.java:75)
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.stop(SuiteTestClassProcessor.java:61)
	... 25 more
Caused by: org.junit.platform.commons.JUnitException: ClassSelector [className = 'com.netflix.nebula.helloworld.HelloReplyOrBuilder'] resolution failed
	at org.junit.platform.launcher.listeners.discovery.AbortOnFailureLauncherDiscoveryListener.selectorProcessed(AbortOnFailureLauncherDiscoveryListener.java:39)
	at org.junit.platform.engine.support.discovery.EngineDiscoveryRequestResolution.resolveCompletely(EngineDiscoveryRequestResolution.java:102)
	at org.junit.platform.engine.support.discovery.EngineDiscoveryRequestResolution.run(EngineDiscoveryRequestResolution.java:82)
	at org.junit.platform.engine.support.discovery.EngineDiscoveryRequestResolver.resolve(EngineDiscoveryRequestResolver.java:113)
	at org.junit.support.testng.engine.TestNGTestEngine.discover(TestNGTestEngine.java:78)
	at org.junit.platform.launcher.core.EngineDiscoveryOrchestrator.discoverEngineRoot(EngineDiscoveryOrchestrator.java:103)
	... 32 more
Caused by: java.lang.NullPointerException
	at org.junit.support.testng.engine.TestAnnotationUtils.getClassHierarchy(TestAnnotationUtils.java:84)
	at org.junit.support.testng.engine.TestAnnotationUtils.collectTestAnnotations(TestAnnotationUtils.java:77)
	at org.junit.support.testng.engine.TestAnnotationUtils.collectGroups(TestAnnotationUtils.java:72)
	at org.junit.support.testng.engine.TestDescriptorFactory.createClassDescriptor(TestDescriptorFactory.java:38)
	at org.junit.support.testng.engine.TestNGSelectorResolver.lambda$resolve$0(TestNGSelectorResolver.java:45)
```

Instead, interfaces and abstract classes should be ignored.

### PR DESCRIPTION
**Pull request by marcphilipp**
Fixes #1.

### CODE REVIEW

File: src/main/java/org/junit/support/testng/engine/TestAnnotationUtils.java [hunk 1]
```diff
@@ -81,7 +81,7 @@ private static Test getAnnotationDirectly(ITestNGMethod method) {
 
 	private static List<Class<?>> getClassHierarchy(Class<?> testClass) {
```
Comment (paplorinc) [inline src/main/java/org/junit/support/testng/engine/TestAnnotationUtils.java:82]: shouldn't this return a stream or iterable instead (based on usage)?

File: src/test/java/org/junit/support/testng/engine/DiscoveryIntegrationTests.java [hunk 2]
```diff
@@ -261,4 +261,20 @@ void ignoresIgnoredTests() {
 		TestDescriptor methodDescriptor = getOnlyElement(classDescriptor.getChildren());
 		assertThat(methodDescriptor.getDisplayName()).isEqualTo("test");
 	}
+
+	@ParameterizedTest
+	@ValueSource(classes = { InterfaceTestCase.class, AbstractTestCase.class })
```
Comment (paplorinc) [inline src/test/java/org/junit/support/testng/engine/DiscoveryIntegrationTests.java:266]: what about enums and records?

Comment (marcphilipp) [reply src/test/java/org/junit/support/testng/engine/DiscoveryIntegrationTests.java]: Verified in e1066c2c8e4bda42a792d907e6fa80e5675d4211.

File: src/main/java/org/junit/support/testng/engine/TestAnnotationUtils.java (hunk 1 above)
Comment (marcphilipp) [reply src/main/java/org/junit/support/testng/engine/TestAnnotationUtils.java]: Thanks for the suggestion! Done in 441ba1bd24f6649c72a36348ee5929c066caf8f3.

File: src/test/java/org/junit/support/testng/engine/DiscoveryIntegrationTests.java [hunk 3]
```diff
@@ -263,7 +263,8 @@ void ignoresIgnoredTests() {
 	}
 
 	@ParameterizedTest
-	@ValueSource(classes = { InterfaceTestCase.class, AbstractTestCase.class })
+	@ValueSource(classes = { InterfaceTestCase.class, AbstractTestCase.class, RecordTestCase.class,
```
Comment (paplorinc) [inline src/test/java/org/junit/support/testng/engine/DiscoveryIntegrationTests.java:266]: would it maybe make sense to try it with JDK 17s https://openjdk.java.net/jeps/409 ?
Not urgent though, just sounds like fun :D

File: src/main/java/org/junit/support/testng/engine/TestAnnotationUtils.java [hunk 4]
```diff
@@ -74,16 +77,27 @@ private static Test getAnnotationDirectly(ITestNGMethod method) {
 	}
 
 	private static Stream<Test> collectTestAnnotations(Class<?> testClass) {
-		return getClassHierarchy(testClass).stream() //
+		return getClassHierarchy(testClass) //
 				.map(clazz -> clazz.getAnnotation(Test.class)) //
 				.filter(Objects::nonNull);
```
Comment (paplorinc) [inline src/main/java/org/junit/support/testng/engine/TestAnnotationUtils.java:82]: when can this be null?

File: src/main/java/org/junit/support/testng/engine/TestAnnotationUtils.java [hunk 5]
```diff
@@ -74,16 +77,27 @@ private static Test getAnnotationDirectly(ITestNGMethod method) {
...
 
-	private static List<Class<?>> getClassHierarchy(Class<?> testClass) {
-		List<Class<?>> result = new ArrayList<>();
-		for (Class<?> clazz = testClass; clazz != Object.class && clazz != null; clazz = clazz.getSuperclass()) {
-			result.add(clazz);
-		}
-		return result;
+	private static Stream<Class<?>> getClassHierarchy(Class<?> testClass) {
+		Iterator<Class<?>> iterator = new Iterator<Class<?>>() {
+			Class<?> next = testClass;
```
Comment (paplorinc) [inline src/main/java/org/junit/support/testng/engine/TestAnnotationUtils.java:87]: should we maybe annotate it as nullable?

File: src/main/java/org/junit/support/testng/engine/TestAnnotationUtils.java [hunk 6]
```diff
@@ -74,16 +77,27 @@ private static Test getAnnotationDirectly(ITestNGMethod method) {
...
+
+			@Override
+			public boolean hasNext() {
+				return next != Object.class && next != null;
+			}
+
+			@Override
+			public Class<?> next() {
+				Class<?> result = next;
+				next = next.getSuperclass();
```
Comment (paplorinc) [inline src/main/java/org/junit/support/testng/engine/TestAnnotationUtils.java:97]: minor: iterators should work without calling `hasNext` - which could NPE here in that case

Comment (paplorinc) [reply src/main/java/org/junit/support/testng/engine/TestAnnotationUtils.java:97]: Scratch that, it's only called by stream.

File: src/main/java/org/junit/support/testng/engine/TestAnnotationUtils.java (hunk 5 above)
Comment (marcphilipp) [reply src/main/java/org/junit/support/testng/engine/TestAnnotationUtils.java:87]: This codebase doesn't use nullable annotations (same for the rest of the JUnit codebase). Could be added but in a separate PR.

File: src/test/java/org/junit/support/testng/engine/DiscoveryIntegrationTests.java (hunk 3 above)
Comment (marcphilipp) [reply src/test/java/org/junit/support/testng/engine/DiscoveryIntegrationTests.java:266]: It shouldn't influence the class hierarchy and the tests are currently compiled with `--release 16`. We can add that when the JDK 17 GA release is out.

**Pull request merged**